### PR TITLE
PSP-15: Batch Mint/Transfer NFT Standard. PSP-16: SubToken Standard.

### PR DESCRIPTION
**Pull request by poria-cat**

### CODE REVIEW

**Comment by Swader** [comment]
Could you also link to accompanying docs and grants so reviewers have context please?

**Review by xlc — COMMENTED** [review]
Rust trait definition isn't really a specification.

The motivation doesn't make mush sense to me and why this needs to be a standard?

**Comment by poria-cat** [comment]
> Could you also link to accompanying docs and grants so reviewers have context please?

Okay, I'm finishing that form.

File: PSPs/drafts/psp-15.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+Tokens get(fn tokens): double_map hasher(blake2_128_concat) T::Hash, hasher(blake2_128_concat) u128 => TokenInfo<T::AccountId>;
+
+/// Use start_idx and end_idx to represent a continuous id,
+/// if end_idx is equal to start_idx, then this is a single NFT, 
+/// if end_idx is greater than start_idx, it is a batch-minted NFT
+pub struct TokenInfo<AccountId> {
+    pub end_idx: u128,
+    pub owner: AccountId,
+    pub uri: Vec<u8>,
+}
```
Comment (semuelle) [inline PSPs/drafts/psp-15.md:29]:
````suggestion
```
/// The set of minted NFTs. (collection_id, start_idx) => nft_info
Tokens get(fn tokens): double_map hasher(blake2_128_concat) T::Hash, hasher(blake2_128_concat) u128 => TokenInfo<T::AccountId>;

/// Use start_idx and end_idx to represent a continuous id,
/// if end_idx is equal to start_idx, then this is a single NFT, 
/// if end_idx is greater than start_idx, it is a batch-minted NFT
pub struct TokenInfo<AccountId> {
    pub end_idx: u128,
    pub owner: AccountId,
    pub uri: Vec<u8>,
}
```
````

File: PSPs/drafts/psp-16.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+Many NFTs are expensive, and if you want more people to have a particular NFT, you can split the NFT so that more people can have it. 
+
+## Specification
+
+/// Stores
+/// The set of SubToken creators. subtoken_collection => creator_address
+SubTokenCreator get(fn sub_token_creator): map hasher(blake2_128_concat) T::Hash => T::AccountId;
+/// Record the collection_id of the SubToken corresponding to the locked NFT subtoken_collection => nft(collection_id, start_idx) 
+SubTokens get(fn sub_tokens): map hasher(blake2_128_concat) T::Hash => (T::Hash, u128);
```
Comment (semuelle) [inline PSPs/drafts/psp-16.md:24]:
````suggestion
```
/// Stores
/// The set of SubToken creators. subtoken_collection => creator_address
SubTokenCreator get(fn sub_token_creator): map hasher(blake2_128_concat) T::Hash => T::AccountId;
/// Record the collection_id of the SubToken corresponding to the locked NFT subtoken_collection => nft(collection_id, start_idx) 
SubTokens get(fn sub_tokens): map hasher(blake2_128_concat) T::Hash => (T::Hash, u128);
```
````

**Review by semuelle — CHANGES_REQUESTED** [review]

File: PSPs/drafts/psp-16.md (hunk 2 above)
Comment (poria-cat) [reply PSPs/drafts/psp-16.md:24]: 😀. Thanks, this looks much better.

**Comment by lamafab** [comment]
> Rust trait definition isn't really a specification.
> 
> The motivation doesn't make mush sense to me and why this needs to be a standard?

I agree with this statement. It's not quite clear why those specific mechanism need to be standardized. We merged a proposal for a [NFT interface](https://github.com/w3f/PSPs/blob/master/PSPs/psp-34.md).

If you want to recommend changes please open an issue/PR for that proposal. I'll close this issue for now.

**Pull request closed without merging**